## Data Preprocessing

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### EDA

In [2]:
df = pd.read_csv("../../ETL/data/fraud_incidents.csv")

In [3]:
df.head(15)

4086aa77-4090-4776-8cb4-2e77ffdf591f  \
0   867c5583-aa80-4318-93fb-302fc35c5cbb   
1   85561331-fb96-446f-bf79-01cae851103a   
2   dcbe13d5-7757-4532-8709-f6347899f6fd   
3   05969593-d0e0-438f-a533-be40ba4fcfbb   
4   759c3d04-a30e-4f56-86db-3899ed74ada4   
5   6b122437-c758-44e8-80c4-07abd69623ad   
6   aab16acc-e9b9-42ee-85f1-8d287a96620e   
7   f222d9e4-ea99-4401-a4ba-dd9081c8e460   
8   98eee8f4-a372-41b4-ae1b-c77b770b304d   
9   66d29759-6448-4eee-8bfb-f30a567d12d2   
10  62966044-3e03-465e-a0a0-a5c5894da037   
11  6b437379-3f52-459b-a5a7-24b80e657036   
12  9ea0bfe4-a45e-4a81-a7a5-a832ca7e3d39   
13  23e4d2d8-2064-4188-a4ef-d8eed22aef14   
14  cab35962-6cb6-4a36-aca4-925b235219b7   

    a315527c-cf1a-4336-b954-1a752fd98482  Unauthorized Transaction  \
0   7b3cd7d7-3bfe-4fe4-b120-58eba10e4d69                  Phishing   
1   1b11ef20-54e3-4e02-ac10-c9dca2a13ade                  Phishing   
2   a9d06942-a250-4d90-89d7-2442bf0f0798            Account Breach   
3   a880d2c2-ddb3-40c8-9ac0-026c18827aff            Account Breach   
4   8a6b3a1e-7723-43b3-a3ab-ee596853d5b3  Unauthorized Transaction   
5   a62e25b6-f970-4555-bf2f-8826e913e58e  Unauthorized Transaction   
6   666b8683-e592-43ad-9c2c-42ca99bfc9be  Unauthorized Transaction   
7   15d1d2ec-1de1-43aa-9cb0-db4a4468f05f            Account Breach   
8   cad74c47-5a10-4e36-811e-2ab581aa0217            Account Breach   
9   0aaba017-7275-46ff-b70b-fdf0492196b6            Account Breach   
10  fe6b7ad3-de2c-4b76-9c02-bbc825b8c84d  Unauthorized Transaction   
11  4e9a520e-544e-444f-bd0e-2cd8952302b5            Account Breach   
12  b9451ac9-7368-4f32-803c-b9c18d39c1bb            Account Breach   
13  4bd64732-69db-4d28-a196-57be761d5019  Unauthorized Transaction   
14  0c90c758-375e-4245-88a3-cf7e876dc6fe                  Phishing   

    2025-03-10 06:55:10        Pending  
0   2025-02-06 15:36:33        Pending  
1   2025-01-27 07:01:10  Investigating  
2   2025-03-07 06:11:22        Pending  
3   2025-02-11 11:22:49        Pending  
4   2025-03-19 23:20:39  Investigating  
5   2025-01-08 15:53:03        Pending  
6   2025-03-21 00:33:59  Investigating  
7   2025-01-18 23:31:10  Investigating  
8   2025-03-05 23:37:09        Pending  
9   2025-03-13 04:35:53        Pending  
10  2025-01-01 04:14:13  Investigating  
11  2025-02-04 11:40:48        Pending  
12  2025-02-11 04:11:18        Pending  
13  2025-03-15 03:32:30  Investigating  
14  2025-01-04 12:51:19        Pending

In [4]:
# 1. Check for missing values
print(df.isnull().sum())

4086aa77-4090-4776-8cb4-2e77ffdf591f    0
a315527c-cf1a-4336-b954-1a752fd98482    0
Unauthorized Transaction                0
2025-03-10 06:55:10                     0
Pending                                 0
dtype: int64


In [5]:
# 2. Check for duplicates
print(df.duplicated().sum())

0


In [6]:
# 3. Check for data types
print(df.dtypes)

4086aa77-4090-4776-8cb4-2e77ffdf591f    object
a315527c-cf1a-4336-b954-1a752fd98482    object
Unauthorized Transaction                object
2025-03-10 06:55:10                     object
Pending                                 object
dtype: object


In [7]:
# 4. Check for unique values
print(df.nunique())

4086aa77-4090-4776-8cb4-2e77ffdf591f    39999
a315527c-cf1a-4336-b954-1a752fd98482    32898
Unauthorized Transaction                    3
2025-03-10 06:55:10                     39890
Pending                                     3
dtype: int64


In [9]:
# 5. Check for the distribution of numerical columns
df.describe()

4086aa77-4090-4776-8cb4-2e77ffdf591f  \
count                                  39999   
unique                                 39999   
top     867c5583-aa80-4318-93fb-302fc35c5cbb   
freq                                       1   

        a315527c-cf1a-4336-b954-1a752fd98482 Unauthorized Transaction  \
count                                  39999                    39999   
unique                                 32898                        3   
top     c8e7babb-5a65-4b3a-85ac-8660da605151           Account Breach   
freq                                       5                    13395   

        2025-03-10 06:55:10  Pending  
count                 39999    39999  
unique                39890        3  
top     2025-03-17 02:01:33  Pending  
freq                      2    13370

In [16]:
df['created_at'] = pd.to_datetime(df['created_at'], errors='coerce')

# Verify changes
print(df.dtypes)

account_id                 object
customer_id                object
account_type               object
account_balance           float64
currency                   object
created_at         datetime64[ns]
dtype: object


### Handle Outliers (IQR Method)

In [13]:
Q1 = df['account_balance'].quantile(0.25)
Q3 = df['account_balance'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
print(lower_bound, upper_bound)


-50091.94625000001 150151.48375


### Your account_balance column has no outliers based on the IQR method. The data is well-distributed.
Explanation: 
Negative Lower Bound (-50091.95): A negative balance makes no sense for a typical account balance if it's not supposed to allow overdrafts.
However, it being negative indicates that there are no negative values in your data — all positive balances fall above the lower bound.

No outliers on the lower side.

Upper Bound (150151.48):The maximum value in your data is 99,999.42, which is well below the upper bound.

No outliers on the upper side.

In [12]:
# Check numerical columns
num_cols = df.select_dtypes(include=[np.number]).columns
print("Numerical columns:", num_cols)


Numerical columns: Index(['account_balance'], dtype='object')


In [ ]:
###